zebROS_ws/src/controllers_2023/src/four_bar_controller.cpp

```diff
@@ -0,0 +1,461 @@
+    }
+
+    ddr_.registerVariable<double>
+    ("parallel_bar_length",
+     [this]()
+    {
+        return parallel_bar_length_.load();
+    },
+    [this](double b)
+    {
+        parallel_bar_length_.store(b);
+    },
+    "Parallel bar length");
+
+    ddr_.registerVariable<double>
+    ("diagonal_bar_length",
+     [this]()
+    {
+        return diagonal_bar_length_.load();
+    },
+    [this](double b)
+    {
+        diagonal_bar_length_.store(b);
+    },
+    "Diagonal bar length");
+
+    ddr_.registerVariable<double>
+    ("intake_length",
+     [this]()
+    {
+        return intake_length_.load();
+    },
+    [this](double b)
+    {
+        intake_length_.store(b);
+    },
+    "Intake/static attachment length");
+
+    ddr_.registerVariable<double>
+    ("arb_feed_forward",
+     [this]()
+    {
+        return arb_feed_forward.load();
+    },
+    [this](double b)
+    {
+        arb_feed_forward.store(b);
+    },
+    "Arb feedforward");
+    ddr_.registerVariable<double>
+    ("four_bar_zeroing_percent_output",
+     [this]()
+    {
+        return four_bar_zeroing_percent_output.load();
+    },
+    [this](double b)
+    {
+        four_bar_zeroing_percent_output.store(b);
+    },
+    "FourBar Zeroing Percent Output");
+    ddr_.registerVariable<double>
+    ("four_bar_zeroing_timeout",
+     [this]()
+    {
+        return four_bar_zeroing_timeout.load();
+    },
+    [this](double b)
+    {
+        four_bar_zeroing_timeout.store(b);
+    },
+    "FourBar Zeroing Timeout");
+    ddr_.registerVariable<double>
+    ("motion_magic_velocity",
+     [this]()
+    {
+        return motion_magic_velocity.load();
+    },
+    [this](double b)
+    {
+        motion_magic_velocity.store(b);
+    },
+    "Motion Magic Velocity");
+    ddr_.registerVariable<double>
+    ("motion_magic_acceleration",
+     [this]()
+    {
+        return motion_magic_acceleration.load();
+    },
+    [this](double b)
+    {
+        motion_magic_acceleration.store(b);
+    },
+    "Motion Magic Acceleration");
+    ddr_.registerVariable<int>
+    ("motion_s_curve_strength",
+     [this]()
+    {
+        return motion_s_curve_strength.load();
+    },
+    [this](int b)
+    {
+        motion_s_curve_strength.store(b);
+    },
+    "S Curve Strength");
+    
+    ddr_.publishServicesTopics();
+
+    //initialize the four_bar joint
+    if (!four_bar_joint_.initWithNode(talon_command_iface, nullptr, controller_nh, four_bar_params))
+    {
+        ROS_ERROR("Cannot initialize four_bar joint!");
+    }
+
+    four_bar_service_ = controller_nh.advertiseService("four_bar_service", &FourBarController_2023::cmdService, this);
+
+    return true;
+}
+
+void FourBarController_2023::starting(const ros::Time &time)
+{
+    zeroed_ = false;
+    last_zeroed_  = false;
+    last_mode_ = hardware_interface::TalonMode_Disabled;
+    last_angle_ = -1; // give nonsense position to force update on first time through update()
```

Comment:
Can't valid angles be negative? Might want to make a it something like a `std::numeric_limits<double>::max();`

Reply:
i think the minimum is 0 because it's relative to min_extension_